crates/neon/src/types_impl/extract/mod.rs

```diff
@@ -0,0 +1,229 @@
+    handle::Handle,
+    result::NeonResult,
+    types::JsValue,
+};
+
+#[cfg(feature = "serde")]
+#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
+pub use self::json::*;
+pub use self::types::*;
+
+#[cfg(feature = "serde")]
+mod json;
+mod types;
+
+/// Extract Rust data from a JavaScript value
+pub trait TryFromJs<'cx>
+where
+    Self: private::Sealed + Sized,
+{
+    type Error;
+
+    fn try_from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Result<Self, Self::Error>>
+    where
+        C: Context<'cx>;
+
+    fn from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Self>
+    where
+        C: Context<'cx>;
+}
+
+/// Trait specifying values that may be extracted from function arguments.
+///
+/// **Note:** This trait is implemented for tuples of up to 32 values, but for
+/// the sake of brevity, only tuples up to size 8 are shown in this documentation.
+pub trait FromArgs<'cx>: private::FromArgsInternal<'cx> {}
+
+// Convenience implementation for single arguments instead of needing a single element tuple
+impl<'cx, T> private::FromArgsInternal<'cx> for T
+where
+    T: TryFromJs<'cx>,
+{
+    fn from_args(cx: &mut FunctionContext<'cx>) -> NeonResult<Self> {
+        let (v,) = private::FromArgsInternal::from_args(cx)?;
+
+        Ok(v)
+    }
+
+    fn from_args_opt(cx: &mut FunctionContext<'cx>) -> NeonResult<Option<Self>> {
+        if let Some((v,)) = private::FromArgsInternal::from_args_opt(cx)? {
+            Ok(Some(v))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl<'cx, T> FromArgs<'cx> for T where T: TryFromJs<'cx> {}
+
+macro_rules! from_args_impl {
+    ($(#[$attrs:meta])? [$($ty:ident),*]) => {
+        $(#[$attrs])?
+        impl<'cx, $($ty,)*> FromArgs<'cx> for ($($ty,)*)
+        where
+            $($ty: TryFromJs<'cx>,)*
+        {}
+
+        #[allow(non_snake_case)]
+        impl<'cx, $($ty,)*> private::FromArgsInternal<'cx> for ($($ty,)*)
+        where
+            $($ty: TryFromJs<'cx>,)*
+        {
+            fn from_args(cx: &mut FunctionContext<'cx>) -> NeonResult<Self> {
+                let [$($ty,)*] = cx.argv();
+
+                Ok(($($ty::from_js(cx, $ty)?,)*))
+            }
+
+            fn from_args_opt(cx: &mut FunctionContext<'cx>) -> NeonResult<Option<Self>> {
+                let [$($ty,)*] = cx.argv();
+
+                Ok(Some((
+                    $(match $ty::try_from_js(cx, $ty)? {
+                        Ok(v) => v,
+                        Err(_) => return Ok(None),
+                    },)*
+                )))
+            }
+        }
+    }
+}
+
+macro_rules! from_args_expand {
+    ($(#[$attrs:meta])? [$($head:ident),*], []) => {};
+
+    ($(#[$attrs:meta])? [$($head:ident),*], [$cur:ident $(, $tail:ident)*]) => {
+        from_args_impl!($(#[$attrs])? [$($head,)* $cur]);
+        from_args_expand!($(#[$attrs])? [$($head,)* $cur], [$($tail),*]);
+    };
+}
+
+macro_rules! from_args {
+    ([$($show:ident),*], [$($hide:ident),*]) => {
```

Comment:
A comment here would be useful to explain that the first list is the ones with docs and the second list is `doc(hidden)`.